python/mpc_cvxgen/mpc.py

```diff
@@ -0,0 +1,321 @@
+Its author is Atsushi Sakai.
+
+Reference Papers:
+- [Kinematic and Dynamic Vehicle Models for Autonomous Driving Control Design]
+(https://borrelli.me.berkeley.edu/pdfpub/IV_KinematicMPC_jason.pdf)
+"""
+
+import numpy as np
+import cvxpy
+from cvxpy.expressions import constants
+from .utils import compute_curvature, Vehicle, Trajectory
+
+
+class ModelPredictiveController:
+    def __init__(self, config):
+        self.reference = Trajectory(**config['reference'])
+        self.vehicle = Vehicle(config['vehicle'])
+        self.path_length = len(self.reference.s_list)
+        self.path_index = 0
+        self.t = 0.0  # [s]
+        initial_condition = {
+            't_list': [self.t],  # Initial time [s]
+            's_list': self.reference.s_list[0:1],  # Initial arc distance [m]
+            'x_list': self.reference.x_list[0:1],  # Initial X coordinate [m]
+            'y_list': self.reference.y_list[0:1],  # Initial Y coordinate [m]
+            'k_list': self.reference.k_list[0:1],  # Initial curvature [1/m]
+            'vel_list': self.reference.vel_list[0:1],  # Initial velocity [m/s]
+            'yaw_list': self.reference.yaw_list[0:1],  # Initial yaw [rad]
+            'accel_list': np.asarray([]),  # Initial acceleration [m/s2]
+            'steer_list': np.asarray([]),  # Initial steering [rad]
+        }
+        self.solution = Trajectory(**initial_condition)
+        self.vehicle.update(
+            self.t,
+            self.solution.s_list[-1],
+            self.solution.x_list[-1],
+            self.solution.y_list[-1],
+            self.solution.k_list[-1],
+            self.solution.vel_list[-1],
+            self.solution.yaw_list[-1],
+            None,
+            None,
+        )
+        self.config = config['controller']
+        self.num_state = 4  # Number of states: X = [x, y, vel, yaw]
+        self.num_input = 2  # Number of inputs: U = [accel, steer]
+        self.position = self.vehicle.get_position()
+        self.goal = np.asarray(
+            [self.reference.x_list[-1], self.reference.y_list[-1]])
+        self.delta_t = 0.1  # [s]
+        self.delta_s = self.reference.s_list[1] - self.reference.s_list[0]
+        self.horizon_accel = np.zeros((self.config['horizon'], 1))  # [m/s2]
+        self.horizon_steer = np.zeros((self.config['horizon'], 1))  # [rad]
+
+    def step(self):
+        """
+        Solve the mpc problem and step the vehicle forward in time with solved
+        controls.
+
+        :return: None
+        """
+        self.t = self.t + self.delta_t
+        # Solve control for the next step
+        self._update_path_index()
+        reference_state, reference_steer = \
+            self._retrieve_imminent_reference()
+
+        for _ in range(self.config['max_iteration']):
+            is_converged = self._iterative_control(reference_state,
+                                                   reference_steer)
+            if is_converged:
```

Comment:
Is there a warning if MPC does not converge? Would be useful to have this as a variable / return it